tests/testsuite/cargo_features.rs

```diff
@@ -113,6 +113,204 @@ release and is no longer necessary to be listed in the manifest
         .run();
 }
 
+#[cargo_test]
+fn allow_features() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                cargo-features = ["test-dummy-unstable"]
+
+                [package]
+                name = "a"
+                version = "0.0.1"
+                authors = []
+                im-a-teapot = true
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    // NOTE: We need to use RUSTC_BOOTSTRAP here since we also need nightly rustc
+
+    p.cargo("-Zallow-features=test-dummy-unstable build")
+        .masquerade_as_nightly_cargo()
+        .env("RUSTC_BOOTSTRAP", "1")
+        .with_stderr(
+            "\
+[COMPILING] a [..]
+[FINISHED] [..]
+",
+        )
+        .run();
+
+    p.cargo("-Zallow-features=test-dummy-unstable,print-im-a-teapot -Zprint-im-a-teapot build")
+        .masquerade_as_nightly_cargo()
+        .env("RUSTC_BOOTSTRAP", "1")
+        .with_stdout("im-a-teapot = true")
+        .with_stderr("[FINISHED] [..]")
+        .run();
+
+    p.cargo("-Zallow-features=test-dummy-unstable -Zprint-im-a-teapot build")
+        .masquerade_as_nightly_cargo()
+        .env("RUSTC_BOOTSTRAP", "1")
+        .with_status(101)
+        .with_stderr(
+            "\
+error: the feature `print-im-a-teapot` is not in the list of allowed features: {\"test-dummy-unstable\"}
+",
+        )
+        .run();
+
+    p.cargo("-Zallow-features= build")
+        .masquerade_as_nightly_cargo()
+        .env("RUSTC_BOOTSTRAP", "1")
+        .with_status(101)
+        .with_stderr(
+            "\
+error: failed to parse manifest at `[..]`
+
+Caused by:
+  the feature `test-dummy-unstable` is not in the list of allowed features: {}
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn allow_features_to_rustc() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "a"
+                version = "0.0.1"
+                authors = []
+            "#,
+        )
+        .file(
+            "src/lib.rs",
+            r#"
+                #![feature(test_2018_feature)]
+            "#,
+        )
+        .build();
+
+    // NOTE: We need to use RUSTC_BOOTSTRAP here since we also need nightly rustc
+
+    p.cargo("-Zallow-features= build")
+        .masquerade_as_nightly_cargo()
+        .env("RUSTC_BOOTSTRAP", "1")
+        .with_status(101)
+        .with_stderr_contains(
+            "\
+[COMPILING] a [..]
+error[E0725]: the feature `test_2018_feature` is not in the list of allowed features
+",
+        )
+        .with_stderr_contains(
+            "\
+error: could not compile `a`
+",
+        )
```

Comment:
Can this be simplified a bit to something like `with_stderr_contains("[..]E0725[..]")`.  If Cargo's testsuite tries to match rustc's output too closely, then it can be a pain when someone tries to make even the slightest change (which requires a coordinated update with cargo).  Also, this doesn't need to check the other output, I think just checking `with_status(101)` is sufficient.